torch_xla/experimental/custom_kernel.py

```diff
@@ -160,47 +160,184 @@ def wrapped_kernel(kernel: Callable,
+  where we only takes q, k, v and causal as input and set block_sizes for the users.
+  """
+
+  MIN_BLOCK_SIZE = 128
+  DEFAULT_MASK_VALUE = -0.7 * float(torch.finfo(torch.float32).max)
+  # The block_sizes configuration is copied from https://github.com/google/maxtext/blob/0fee320451738166c8e596dc63a57a4673671576/MaxText/layers/attentions.py#L215-L240
+  # It yields much better performance than the default block_sizes.
+  DEFAULT_BLOCK_SIZES = {
+      "block_q": 512,
+      "block_k_major": 512,
+      "block_k": 512,
+      "block_b": 2,
+      "block_q_major_dkv": 512,
+      "block_k_major_dkv": 512,
+      "block_q_dkv": 512,
+      "block_k_dkv": 512,
+      "block_q_dq": 1024,
+      "block_k_dq": 256,
+      "block_k_major_dq": 512,
+  }
+
+  @staticmethod
+  def forward(ctx, q, k, v, causal=False):
+    # Import JAX within the function such that we don't need to call the jax_import_guard()
+    # in the global scope which could cause problems for xmp.spawn.
+    jax_import_guard()
+    from jax.experimental.pallas.ops.tpu.flash_attention import _flash_attention_impl
+
+    ctx.causal = causal
+    save_residuals = q.requires_grad or k.requires_grad or v.requires_grad
+
+    # It returns the shape and type of o, l, m.
+    def shape_dtype(q, *arg):
+      if not save_residuals:
+        return [(q.shape, q.dtype)]
+      res_shape = list(q.shape)
+      res_shape[-1] = FlashAttention.MIN_BLOCK_SIZE
+      return [(q.shape, q.dtype), (res_shape, torch.float32),
+              (res_shape, torch.float32)]
+
+    # We can't directly use flash_attention as we need to override the save_residuals flag which returns
+    # l and m that is needed for the backward. Then we lose all the shape checks.
+    # TODO: replicate the shape checks on flash_attention.
+    _flash_attention_impl = make_kernel_from_pallas(_flash_attention_impl,
+                                                    shape_dtype)
+    with torch.no_grad():
+      o = _flash_attention_impl(
+          q,
+          k,
+          v,
+          None,
+          None,
+          save_residuals,
+          causal,
+          1.0,
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_b"], q.shape[0]),
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_q"], q.shape[2]),
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_k_major"], k.shape[2]),
+          min(FlashAttention.DEFAULT_BLOCK_SIZES["block_k"], k.shape[2]),
+          False,
+          static_argnums=range(5, 13))
+      if not save_residuals:
+        return o
+      o, *aux = o
+      l, m = (v[..., 0] for v in aux[-2:])
+
+    ctx.save_for_backward(q, k, v, o, l, m)
+    return o
+
+  @staticmethod
+  def backward(ctx, grad_output):
+    # Import JAX within the function such that we don't need to call the jax_import_guard()
+    # in the global scope which could cause problems for xmp.spawn.
+    jax_import_guard()
```

Comment:
we shouldn't need this right? The forward should have been called at this point and fwd/bwd happens in the same process.

Reply:
Right!